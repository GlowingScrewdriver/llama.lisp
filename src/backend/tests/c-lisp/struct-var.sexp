(c-lisp
    (define-struct Node
        (i int)
        (f float))

    (define ((print int) (n int)))
    (define ((fprint float) (n float)))

    (define ((node-print void) (node (struct Node)))
        (call print (load (member-ref node i)))
        (call fprint (load (member-ref node f)))
        (ret))

    (define ((struct-init void) (node (ptr (struct Node))))
        (declare (val (struct Node)))
        (store (member-ref val i) 4)
        (store (member-ref val f) 5.0)

        (store node val)
        (ret))

    (define ((main void))
        (declare (n-ptr (ptr (struct Node))))
        (set n-ptr (alloc (struct Node) 1))
        (call struct-init n-ptr)

        (declare (n-var (struct Node)))
        (set n-var (load n-ptr))
        (call node-print n-var)
        (ret)))
