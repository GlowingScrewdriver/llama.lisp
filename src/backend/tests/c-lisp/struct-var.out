4
5.000000
